### PR TITLE
Make sequence unique

### PR DESCRIPTION
**Pull request by bzwei**
Remove dependency on acts_as_list. Implement a simple logic for sequence.

https://projects.engineering.redhat.com/browse/SSP-1471

### CODE REVIEW

**Comment by gmcculloug** [comment]
@bzwei What happens when you delete one?  Will it create gaps that never get filled in?  Since the number is visible that seems like an issue.

**Comment by bzwei** [comment]
@gmcculloug Gap is allowed even with `acts_as_list` that we used before. What guaranteed is the sequence order. 
Not only the delete method creates a gap, the adjustment of sequence also generates a gap. I think it is OK. We could try to avoid gaps by conducting more `update_all` calls with a tradeoff of performance.

File: app/models/workflow.rb [hunk 1]
```diff
@@ -18,4 +21,24 @@ def external_processing?
...
+
+    self.sequence = self.class.last&.sequence.to_i + 1
+  end
+
+  # move all related sequence number one higher if the desired number is in use
+  def move_rest_higher
+    return unless sequence && self.class.where(:sequence => sequence).exists
+
+    self.class.where(table[:sequence].gteq(sequence)).update_all("sequence = (-sequence - 1)")
+    self.class.where(table[:sequence].lt(0)).update_all("sequence = (-sequence)")
```
Comment (hsong-rh) [inline app/models/workflow.rb:42]: should these two steps are `locked` to prevent concurrent issue?

Comment (bzwei) [reply app/models/workflow.rb:42]: try to avoid locking by using transaction and retry.

Comment (hsong-rh) [reply app/models/workflow.rb:42]: I saw it in workflow `update` only, but no in `create`, any reason?

Comment (hsong-rh) [reply app/models/workflow.rb:42]: I saw it in workflow `update` only, but no in `create`, any reason?

File: app/models/workflow.rb [hunk 2]
```diff
@@ -18,4 +21,24 @@ def external_processing?
...
   end
+
+  private
+
+  def table
+    self.class.arel_table
+  end
+
+  def new_sequence
+    return move_rest_higher if sequence
```
[the remark this answers is not in the document]
Comment (bzwei) [reply app/models/workflow.rb:32]: @mkanoor sequence will be auto generated if it is not given in the attribute list when a new workflow is created. However the sequence will be honored if it is given and thus need to push other sequences higher if there is a conflict.

File: app/models/workflow.rb (hunk 1 above)
Comment (bzwei) [reply app/models/workflow.rb:42]: create by default will always place the new workflow to the end of the sequencing, unless the number is given by caller and this method does get called.

BTW, transaction is not needed in our service classes. See the document https://guides.rubyonrails.org/active_record_callbacks.html#halting-execution. I am going to remove them.

Comment (hsong-rh) [reply app/models/workflow.rb:42]: 👍

**Pull request merged**